### ISSUE TITLE
[Feature Request] Implement FastCDC?

### ISSUE THREAD

**Issue by reisub0** [librsync/librsync#242]
Google [recently open-sourced](https://github.com/google/cdc-file-transfer) an implementation for [FastCDC](https://www.usenix.org/conference/atc16/technical-sessions/presentation/xia) (specifically geared towards Linux-Windows transfers of games). 

Across their tests and benchmarks, CDC [outperformed the rsync protocol](https://github.com/google/cdc-file-transfer#how-does-it-work-and-why-is-it-faster) (in Linux-Linux syncs)

> Because of this, the cdc_rsync algorithm is faster than the standard rsync. It is also simpler. Since chunk boundaries move along with insertions or deletions, the task to match local and remote hashes is a trivial set difference operation. It does not involve a per-byte hash map lookup.

This naturally leads to the following question:

Is it possible to provide CDC in the core rsync algorithm, and what would this involve?

**Comment by dbaarda** [comment]
I've actually done a bit of analysis of FastCDC and chunkers in general [here](https://github.com/dbaarda/rollsum-chunking/blob/master/RESULTS.rst). The interesting result was that FastCDC is not actually better than a simple exponential chunker provided you select the right parameters. I might try submitting a pull request.

It's true that CDC will be faster than rsync because the hashtable lookups are expensive, and CDC avoids that. I suspect that rsync will give better deduplication, because CDC depends on the chunk-boundaries re-syncing after a delta and this can skip over matches that rsync would find. There might also be benefits to fixed size blocks for some applications.

I guess it could be done, but I'm not sure if librsync is the best place to do it. The implementation details are pretty different. It might make sense to do it in a completely different library.

**Comment by ljusten** [comment]
We used librsync before FastCDC in [cdc-file-transfer](https://github.com/google/cdc-file-transfer), and in my experience the dedupe rates of CDC were marginally better (~1% maybe), but the difference is probably not significant. I think the reason was that CDC can use smaller chunk sizes. We use 8K, whereas librsync selects larger chunks of ~30 KB for ~1G files IIRC. That's another advantage of CDC: The chunk size only has a minimal effect on the performance of the algorithm. The only practical lower limit is how fast you can send hashes from remote to local. For rsync, small chunk sizes mean larger hash table lookups, which might trash the cache.

**Comment by dbaarda** [comment]
Actually, I've been playing with cdc_file_transfer and figure it's probably worth implementing in librsync. If nothing else it will give us an apples-to-apples comparison for the algorithms and a common cmdline tool.

**Comment by dbaarda** [comment]
The good news is the delta file format will work for dynamic-sized blocks, so it doesn't need to change, and thus the patch will work unchanged. This means we need a new signature format and generator, and a new delta generator for that signature type. I think I'll leverage off the stuff I started playing with in #209 which gives a more plug-able callback API.

I might also add a Gear rollsum alternative. It's a pretty terrible rollsum for rsync because it only really includes the last 32 bytes in the checksum, but it's a good one for CDC because it only wants a small window and it's fast. It might also be interesting to see how terrible it would be for rsync.

**Comment by dbaarda** [comment]
No, I haven't added it yet. I did contribute some CDC improvements to [cdc-file-transfer](https://github.com/google/cdc-file-transfer/pull/79) based on research I'd done on this. After I do a librsync release tonight I'll start working on adding CDC support to librsync.

**Comment by polar1shu** [comment]
Well. I'm looking forward to its performance.
